File: 00xx/0046-permutations/46_1.ts
/* 46. Permutations */

/* Complexities.
 * Time: O(n n!)
 * Space: O(n)
 * where n is the length of the input array. */

function permute(nums: number[]): number[][] {
  if (nums.length == 0) {
    return [[]];
  }

  let result: number[][] = [];

  for (const x of nums) {
    /* Filter is somewhat overkill, but LeetCode doesn’t support
     * structuredClone + splice, nor toSpliced. */
    const recResult = permute(nums.filter((y) => y !== x));
    recResult.forEach((result) => result.push(x));
    result = result.concat(recResult);
  }

  return result;
}

File: 04xx/0461-hamming-distance/461.c
/* 461. Hamming Distance */

/* Complexities.
 * Time: Θ(d)
 * Space: Θ(1)
 * where d is the Hamming distance of the two input numbers. */

int hammingDistance(int x, int y);
int countBits(int x);

int hammingDistance(int x, int y)
{
	return countBits(x ^ y);
}

int countBits(int x)
{
	int counter = 0;
	while (x != 0) {
		++counter;
		x &= x - 1;
	}
	return counter;
}
